cmd/cstor-pool-mgmt/controller/pool-controller/new_pool_controller.go

```diff
@@ -0,0 +1,144 @@
+	// time, and makes it easy to ensure we are never processing the same item
+	// simultaneously in two different workers.
+	workqueue workqueue.RateLimitingInterface
+	// recorder is an event recorder for recording Event resources to the
+	// Kubernetes API.
+	recorder record.EventRecorder
+}
+
+// NewCStorPoolController returns a new instance of CStorPool controller
+func NewCStorPoolController(
+	kubeclientset kubernetes.Interface,
+	clientset clientset.Interface,
+	kubeInformerFactory kubeinformers.SharedInformerFactory,
+	cStorInformerFactory informers.SharedInformerFactory) *CStorPoolController {
+
+	// obtain references to shared index informers for the cStorPool resources
+	cStorPoolInformer := cStorInformerFactory.Openebs().V1alpha1().CStorPools()
+
+	openebsScheme.AddToScheme(scheme.Scheme)
+
+	// Create event broadcaster to receive events and send them to any EventSink, watcher, or log.
+	// Add NewCstorPoolController types to the default Kubernetes Scheme so Events can be
+	// logged for CstorPool Controller types.
+	glog.V(4).Info("Creating event broadcaster")
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+
+	// StartEventWatcher starts sending events received from this EventBroadcaster to the given
+	// event handler function. The return value can be ignored or used to stop recording, if
+	// desired. Events("") denotes empty namespace
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeclientset.CoreV1().Events("")})
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: poolControllerName})
+
+	controller := &CStorPoolController{
+		kubeclientset: kubeclientset,
+		clientset:     clientset,
+		deletedIndexer: cache.NewIndexer(cache.DeletionHandlingMetaNamespaceKeyFunc,
+			cache.Indexers{}),
+		cStorPoolSynced: cStorPoolInformer.Informer().HasSynced,
+		workqueue:       workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "CStorPool"),
+		recorder:        recorder,
+	}
+
+	glog.Info("Setting up event handlers")
+
+	// Instantiating QueueLoad before entering workqueue.
+	q := common.QueueLoad{}
+
+	// Set up an event handler for when CstorPool resources change.
+	cStorPoolInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			if !IsRightCStorPoolMgmt(obj.(*apis.CStorPool)) {
+				return
+			}
+			q.Operation = "add"
+			glog.Infof("added event")
+			controller.enqueueCStorPool(obj, q)
+		},
+		UpdateFunc: func(old, new interface{}) {
+			newCStorPool := new.(*apis.CStorPool)
+			oldCStorPool := old.(*apis.CStorPool)
+			// Periodic resync will send update events for all known CStorPool.
+			// Two different versions of the same CStorPool will always have different RVs.
+			if newCStorPool.ResourceVersion == oldCStorPool.ResourceVersion {
+				return
+			}
+			if !IsRightCStorPoolMgmt(newCStorPool) {
```

Comment:
This method should be the first method in case handler.

Reply:
we will have to decide based on frequency and sharedinformer interval.

Reply:
I think shared watched interval is fixed and seems to be more frequent and we can have fake update first.